IPython/kernel/comm/comm.py

```diff
@@ -63,6 +66,10 @@ def __init__(self, target_name='', data=None, **kwargs):
     
     def _publish_msg(self, msg_type, data=None, metadata=None, buffers=None, **keys):
         """Helper for sending a comm message on IOPub"""
+        if threading.current_thread().name != 'MainThread' and IOLoop.initialized():
+            # make sure we never send on a zmq socket outside the main IOLoop thread
+            IOLoop.instance().add_callback(lambda : self._publish_msg(msg_type, data, metadata, buffers, **keys))
```

Comment:
Ah, gotcha. Yeah, it would be easy for callbacks that register callbacks to result in starvation, so IOLoop is careful to avoid that.